Review for first release of mangopy

ljlamarche opened the pull request:
This is for the first release of mangopy.  Please take a look and provide suggestions for any/all aspects of it (code, documentation, package installation, style, ect.)!

ljlamarche committed:
Added initial version of mosaic.py

ljlamarche committed:
Added create_hdf5_files.py which generates a single hdf5 file from all the image files for a single day.

ljlamarche committed:
Added Latitude/Longitude arrays to *.h5 file created by create_h5_files.py. Modifed mosaic.py to read in new format of *.h5 files.

astib committed:
test commit by astib

ljlamarche committed:
Modified mosaic.py so that it does nearest-neighbor interpolation based on regrid*.txt look-up files.

ljlamarche committed:
Optimized regridding procedure for mosaic.py.

ljlamarche committed:
Save regrid arrays as single hdf5 file instead of *.txt files.

ljlamarche committed:
Added mango.py which has basic data reading and plotting functions for a single camera.

ljlamarche committed:
Removed hard-coded file paths from mango.py and added SiteInformation.csv.

ljlamarche committed:
Modified mango.py so site is passed to methods insted of an attribute of the class.

ljlamarche committed:
Made mosaic.py use data_read() method from mango.py.

ljlamarche committed:
Converted mango.py and mosaic.py into the installable package mangopy.

ljlamarche committed:
Added GNU GPL v3 license.

ljlamarche committed:
Fixed filepaths for SiteInformation.csv and MANGOData/ so that they can be found regardless of where mangopy is run from.

ljlamarche committed:
Modified read_data() so an error is raised if there is no timestamp within 5 minutes of the requested time.

ljlamarche committed:
Added the true time of each image back below the map in plot_mosaic().

ljlamarche committed:
Added optional dpi flag to plot_mosaic() to allow the resolution of the saved output image to be specified.

ljlamarche committed:
Replaced the triple for loop that assigned the common grid based on hiarchy array with advanced indexing (much faster).

ljlamarche committed:
Refactored moasic.py so generating the grid and interoplating data are handled more independently.

ljlamarche committed:
Added funciton create_mosaic_movie() to mosaic.py to generate a *.mp4 for a complete night.

ljlamarche committed:
Fixed bugs in recalculating the nearest_idx array in get_nearest_index().

ljlamarche committed:
Changed default data directory to be outside of the git repository.  Corrected the filepath to regrid_image_index.h5.

ljlamarche committed:
Added installation instructions to README.md and a simple tutorial.

ljlamarche committed:
Fixed datadir init input to Mosaic() class.

ljlamarche committed:
Modified create_hdf5_files.py so that it reads file paths from config.ini and loops over all sites automatically.

ljlamarche committed:
Fixed data fetching utility.

ljlamarche committed:
Cleaned up error handling in data fetching utility.

ljlamarche committed:
Cleaned up namespace so Mango() and Mosaic() classes are directly part of mangopy.

ljlamarche committed:
Added demonstration of getting data arrays to tutorial.

ljlamarche committed:
Changed default MANGO data directory to be defined by tempfile if user does not provide a directory.

asreimer commented on DataProcessing/MANGO_HDF_convert.py:
```diff
@@ -0,0 +1,26 @@
+import matplotlib
```
Recommend adding some docstring to the top of this file explaining its purpose. Even just a comment would be good.

ljlamarche replied:
It looks like MANGO_HDF_convert.py is actually not necessary and not used anymore.  The relevant bits of code were pulled into create_hdf5_files.py when we reformatted the output hdf5 files.  Maybe we can just remove this script from the repo?

asreimer replied:
That's probably a good idea. If it's not needed, might as well prune it.

asreimer commented on DataProcessing/latLonDecoder.py:
```diff
@@ -0,0 +1,82 @@
+#########################################################
+#### takes newi and newj .csvs and converts them to lat long ####
+#### this code is based on undoing a portion of what pixelToLatLon did ####
+from numpy import *
```
This is generally not a good practice, better to just do what was done in the next line: `import numpy as np`.

I think the only things that need to be fixed after removing this line are:
1) replace all "NAN" with "np.nan"
2) replace "float" with "np.float"
3) replace "empty" with "np.empty"
4) replace "genfromtxt" with "np.genfromtxt"
5) remaining changes needed are commented below

ljlamarche replied:
Agreed, corrections made throughout file.

asreimer replied:
Did you push the changes? I'm not seeing them here, but that might be me missing something.

ljlamarche replied:
Nope, sorry.  Changes should be pushed now!

asreimer commented on DataProcessing/latLonDecoder.py:
```diff
@@ -0,0 +1,82 @@
+    for i in range(ImageWidth):
+        Icell = IMatrix[j][i]
+        Jcell = JMatrix[j][i]
+        if (not(isnan(Icell)) and not(isnan(Jcell))):
```
"isnan" should be "np.isnan"

asreimer commented on DataProcessing/latLonDecoder.py:
```diff
@@ -0,0 +1,82 @@
+        if (not(isnan(Icell)) and not(isnan(Jcell))):
+            pixLon = lonBounds[0] + ((Icell*lonRange)/np.float64(newImageWidth -1)) #perfect!
+            scaleJ = scaleBounds[0] + (scaleRange)*((np.float64(newImageHeight-1)-Jcell)/(newImageHeight-1))
+            pixLat = degrees(2*(arctan(exp(scaleJ))-(pi/4))) 
```
add "np." to "degrees", "arctan", "exp", and "pi"

asreimer commented on DataProcessing/latLonDecoder.py:
```diff
@@ -0,0 +1,82 @@
+            lonMatrix[int(Jcell),int(Icell)] = pixLon
+            
+#### calculate the nonnan SW and NE corners of each location
+southWestLatLon = [nanmin(latMatrix), nanmin(lonMatrix)]
```
add "np." to "nanmin"

asreimer commented on DataProcessing/latLonDecoder.py:
```diff
@@ -0,0 +1,82 @@
+imageLatitudesLongitudes = array((vstack((southWestLatLon, northEastLatLon))))  # creates an array stacked for writing to a CSV
+
+#### save the matricis to files in their respctive directories ####
+savetxt('Latitudes.csv', latMatrix, delimiter=',')
```
add "np." to "savetxt"

asreimer commented on README.md:
````diff
@@ -1,3 +1,21 @@
+pip install .
+```
+
+For basic usage examples, please refer to the mangopy tutorial jupyter notebook, `mangopy_tutorial.ipynb`.
````
Could link to "https://github.com/astib/MANGO/blob/master/mangopy_tutorial.ipynb" directly. This URL doesn't exist yet, but will once this pull request is merged.

asreimer commented on mangopy/mosaic.py:
```diff
@@ -0,0 +1,304 @@
+        return np.array([grid_lon, grid_lat]), np.array([edge_lon, edge_lat])
+
+
+    def site_hiarchy(self,grid_points):
```
Intended typo? If not, should be "hierarchy"

ljlamarche replied:
Nope, that's a typo.  Thanks for the catch!  I'll change it throughout.

asreimer commented on mangopy/mosaic.py:
```diff
@@ -0,0 +1,304 @@
+        truetime = []
+        for site in self.site_list:
+
+            print(site['name'])
```
Is this leftover from debugging or wanted output when function is run?

ljlamarche replied:
This is left-over debugging, but it's useful when there are many sites where data does not exist so the data fetching throws exceptions.  Maybe a better solution would be to have the site name specified in the exceptions?  It's fairly common for data to not exist for some sites, but the way this is currently handled kind of creates a "text vomit" to the screen (e.g., last part of tutorial).  Maybe we could create a verbose keyword that specifies whether or not these warnings should be printed?

asreimer replied:
That's a good idea to include the site name in the error.

Not necessary right now, but it might be a good idea to use the `logging` module to log errors and exceptions, etc. This gives the user control over the amount of vomit they see: https://docs.python.org/3/library/logging.html

ljlamarche replied:
That's a good idea.  I'll open an issue on logging errors and exceptions and it can be addressed in a later release.

asreimer commented on mangopy/mosaic.py:
```diff
@@ -0,0 +1,304 @@
+                img, __, __, tt = self.get_data(site,time)
+                truetime.append(tt)
+            except (OSError, IOError, ValueError) as e:
+                print(e)
```
Suggestion: 
print("Exception: %s' % str(e))

ljlamarche replied:
Looks good, that's more informative!

asreimer commented on mangopy/mosaic.py:
```diff
@@ -0,0 +1,304 @@
+        grid, edges = self.generate_grid()
+
+        # find site hiarchy for background grid
+        hiarchy = self.site_hiarchy(grid)
```
Again, possible typo? hierarchy

ljlamarche replied:
Yep, fixed.

asreimer commented on mangopy/mosaic.py:
```diff
@@ -0,0 +1,304 @@
+            print(time)
+
+            # create mosaic of all sites on background grid
+            mosaic, truetime = self.grid_mosaic(time,grid,hiarchy)
```
hierarchy

ljlamarche replied:
Fixed!

asreimer commented on mangopy/mango.py:
```diff
@@ -0,0 +1,187 @@
+        try:
+            img_array, lat, lon, truetime = self.read_datafile(filename,targtime)
+        # if that fails, try to download, then read the data file
+        except OSError:
```
Perhaps one should be verbose about this? Print something out saying that the file needs to be fetched?

ljlamarche replied:
Sounds good!

asreimer commented on mangopy/mango.py:
```diff
@@ -0,0 +1,187 @@
+        self.mangopy_path = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+        # if no data directory specified, use a default temp directory
+        if datadir is None:
+            datadir = os.path.join(tempfile.gettempdir(),'MANGOData')
```
Maybe one should be verbose about the creation of this directory? Print something to let the user know?

ljlamarche replied:
I'm fine printing a statement to tell users what has been chosen as the default data directory.  Do you think it makes more sens to do it here or down at line 109 where the directory is actually created?

asreimer replied:
Whoops, ya, definitely makes more sense to do it where the directory is created.

asreimer commented:
Looks good to me. In addition to the comments in my review, I have a couple more:
1) What versions of python is this code working in?
2) The README could include some general information about how the package works, specifically: mangopy puts mango files in a data directory and if they don't exists there, they are pulled via FTP. 
3) The tutorial is great! But the last line in it has a problem, see: https://github.com/astib/MANGO/blob/develop/mangopy_tutorial.ipynb

ljlamarche committed:
Updated SiteInformation.csv.

ljlamarche committed:
Corrected numpy import in latLonDecoder.py.

ljlamarche commented:
Thanks @asreimer !  To address your questions:
1. I've been testing this in python 3.6, but I'll give it a try in python 2.7.
2. Good point about the README!  I'll add that in.
3. The last line in the tutorial is actually expected (although probably undesirable) behavior.  The root of the issue is that there are often time when not all sites have data available.  Currently, the mosaic code spits out a lot of errors when this happens and skips to the next site, but maybe it's ok to hide some of those exceptions, at least by default.  Maybe we can put in a verbose option?

asreimer commented:
Hmmm. It seems that one of my comments about the `setup.py` didn't make it into the original review. One can use the `install_requires` keyword to specify the dependencies. It is suggested that one does this using a `requirements.txt` file: https://packaging.python.org/discussions/install-requires-vs-requirements/

or even better, a `setup.cfg` file: https://docs.python.org/3/distutils/configfile.html

ljlamarche committed:
Improved Exception messages and tutorial.

ljlamarche committed:
Added requirements for installation.

ljlamarche committed:
Modified hdf5 processing scripts so they are up-to-date with what's on server.

asreimer commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+numpy >= 1.16
+h5py >= 2.9
+matplotlib >= 3.0
```
If this python package is going to be backward compatible with python 2.7, we need to relax this constraint. Matplotlib v3 is not available on python 2.7. Latest version for py27 is matplotlib 2.2.4

asreimer commented:
Everything works great on python 3.7.3.

asreimer commented:
Works perfectly in python 2.7, but only if the matplotlib requirement is changed from `>=3.0` to `2.2.4`.

ljlamarche committed:
Modified requirements.txt and data fetching utility so they are python 2 and python 3 comptatable.

ljlamarche committed:
Improved project description and installation instructions in README.md

ljlamarche commented:
Ok, I modified `fetch_datafile` to use `ftplib` instead of `urllib` (`requests` doesn't actually appear to work for ftp).  I also changed `requirements.txt` as recommended.  Based on my testing in conda python 2 and python 3 environments, all of this works now!

Unless there are any more small outstanding issues, I'm fine merging this PR at this point.

ljlamarche committed:
Reduced traceback for exception raised when data does not exist.

asreimer commented:
Looks good! All my testing shows things working really nicely in python2 and 3.
